roomserver/api/query.go

```diff
@@ -348,6 +348,18 @@ type QueryServerBannedFromRoomResponse struct {
 	Banned bool `json:"banned"`
 }
 
+type QueryRestrictedJoinAllowedRequest struct {
+	UserID string `json:"user_id"`
+	RoomID string `json:"room_id"`
+}
+
+type QueryRestrictedJoinAllowedResponse struct {
+	Restricted    bool   `json:"restricted"`               // Is the room membership restricted?
+	Resident      bool   `json:"resident"`                 // Is our homeserver in the relevant rooms?
+	Allowed       bool   `json:"allowed"`                  // Is the join allowed by the rules?
```

Comment:
All of these flags need additional comments. They aren't common terms in use so it's unclear what they mean. Please explain in simple terms.